renku/core/session/notebook_service.py

```diff
@@ -0,0 +1,381 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Docker based interactive session provider."""
+
+import urllib
+from pathlib import Path
+from time import monotonic, sleep
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from lazy_object_proxy import Proxy
+
+from renku.command.command_builder import inject
+from renku.core import errors
+from renku.core.interface.client_dispatcher import IClientDispatcher
+from renku.core.management.client import LocalClient
+from renku.core.plugin import hookimpl
+from renku.core.session.utils import get_renku_project_name, get_renku_url
+from renku.core.util import communication, requests
+from renku.core.util.git import get_remote
+from renku.domain_model.session import ISessionProvider, Session
+
+
+def _get_token(client: LocalClient, renku_url: str) -> Tuple[str, bool]:
+    """Get a token for authenticating with renku.
+
+    If the user is logged in then the JWT token from renku login will be used.
+    Otherwise the anonymous user token will be used. Returns the token and a flag to
+    indicate if the user is registered (true) or anonymous(false).
+    """
+    registered_token = client.get_value(section="http", key=urllib.parse.urlparse(renku_url).netloc)
+    if not registered_token:
+        return _get_anonymous_credentials(client=client, renku_url=renku_url), False
+    return registered_token, True
+
+
+def _get_anonymous_credentials(client: LocalClient, renku_url: str) -> str:
+    def _get_anonymous_token() -> Optional[str]:
+        import requests
+
+        with requests.Session() as session:
+            url = urllib.parse.urljoin(renku_url, "api/user")
+            try:
+                session.get(
+                    url,
+                    headers={
+                        "X-Requested-With": "XMLHttpRequest",
+                        "X-Forwarded-Uri": "/api/user",
+                    },
+                )
+            except (requests.exceptions.RequestException, requests.exceptions.ConnectionError):
+                pass
+            return session.cookies.get("anon-id")
+
+    renku_host = urllib.parse.urlparse(renku_url).netloc
+    anon_token = client.get_value(section="anonymous_token", key=renku_host)
+    if not anon_token:
+        anon_token = _get_anonymous_token()
+        if not anon_token:
+            raise errors.AuthenticationError(
+                "Could not get anonymous user token from Renku. "
+                f"Ensure the Renku deployment at {renku_url} supports anonymous sessions."
+            )
+        client.set_value(section="anonymous_token", key=renku_host, value=anon_token, global_only=True)
+    return anon_token
+
+
+class NotebookServiceSessionProvider(ISessionProvider):
+    """A session provider that uses the notebook service API to launch sessions."""
+
+    DEFAULT_TIMEOUT_SECONDS = 300
+    _client_dispatcher = None
+
+    @classmethod
+    def client_dispatcher(cls):
```

Comment:
```suggestion
    @property
    def client_dispatcher(self):
```

It being a property should work as well and make code accessing it look a bit cleaner

Reply:
Ah cool, I wasn't sure if there were performance or other concerns in making this a property. I will make it a property then.

Reply:
Ok so `@property` does not work with the injection or some other thing. But I could just make it a method that returns the client. And that works just fine. And I do not mind putting an extra set of brackets when calling the dispatcher.